peer_name_mac.go

```diff
@@ -45,11 +46,35 @@ func PeerNameFromUserInput(userInput string) (PeerName, error) {
 
 // PeerNameFromString parses PeerName from a generic string.
 func PeerNameFromString(nameStr string) (PeerName, error) {
-	mac, err := net.ParseMAC(nameStr)
-	if err != nil {
-		return UnknownPeerName, err
+	var a, b, c, d, e, f uint64
+
+	match := func(format string, args ...interface{}) bool {
+		a, b, c, d, e, f = 0, 0, 0, 0, 0, 0
+		n, err := fmt.Sscanf(nameStr+"\000", format+"\000", args...)
+		return err == nil && n == len(args)
+	}
+
+	switch {
+	case match("%2x:%2x:%2x:%2x:%2x:%2x", &a, &b, &c, &d, &e, &f):
+	case match("::%2x:%2x:%2x:%2x", &c, &d, &e, &f):
+	case match("%2x::%2x:%2x:%2x", &a, &d, &e, &f):
+	case match("%2x:%2x::%2x:%2x", &a, &b, &e, &f):
+	case match("%2x:%2x:%2x::%2x", &a, &b, &c, &f):
+	case match("%2x:%2x:%2x:%2x::", &a, &b, &c, &d):
+	case match("::%2x:%2x:%2x", &d, &e, &f):
+	case match("%2x::%2x:%2x", &a, &e, &f):
+	case match("%2x:%2x::%2x", &a, &b, &f):
+	case match("%2x:%2x:%2x::", &a, &b, &c):
+	case match("::%2x:%2x", &e, &f):
+	case match("%2x::%2x", &a, &f):
+	case match("%2x:%2x::", &a, &b):
+	case match("::%2x", &f):
+	case match("%2x::", &a):
+	default:
+		return UnknownPeerName, fmt.Errorf("Invalid peer name format")
```

Comment:
> Error strings should not be capitalized (unless beginning with proper nouns or acronyms) or end with punctuation, since they are usually printed following other context. That is, use fmt.Errorf("something bad") not fmt.Errorf("Something bad")

https://github.com/golang/go/wiki/CodeReviewComments#error-strings